### PR TITLE
Use errors.Is() in IsErrEOF()

### PR DESCRIPTION
**Pull request by pierreca**
IsErrEOF returns false when it should return true in a couple of cases:

1. if the error has been wrapped in another error (for example, if EOF
is wrapped in an RPC error)
2. if the error has been created from an Error field in an RPC response
(as it is the case in CallWithCodec in the net-rpc-msgpackrpc package
for example)

### CODE REVIEW

**Review by platttammy — APPROVED** [review]

File: lib/eof.go [hunk 1]
```diff
@@ -13,13 +14,14 @@ var yamuxSessionShutdown = yamux.ErrSessionShutdown.Error()
...
-	if err == io.EOF {
+	if errors.Is(err, io.EOF) {
 		return true
 	}
 
 	errStr := err.Error()
 	if strings.Contains(errStr, yamuxStreamClosed) ||
-		strings.Contains(errStr, yamuxSessionShutdown) {
+		strings.Contains(errStr, yamuxSessionShutdown) ||
+		strings.HasSuffix(errStr, io.EOF.Error()) {
```
Comment (dnephin) [inline lib/eof.go:24]: In theory this seems reasonable, but I think because `EOF.Error()` is such a short string (`EOF`), there is a pretty significant chance this incorrectly matches a a different error.

A suffix of `: EOF` might be a bit safer, but I'm not sure that is necessary either. I think we should be encouraging all code to use wrapping and not rely on string comparison. Ideally we would remove the need for these other `strings.Contains()` as well.

I think we should not add this `strings.HasSuffix`, and fix any code that does not properly wrap errors to do so.

**Review by dnephin — COMMENTED** [review]
Thank you for this PR! 

I think using `errors.Is` is the right thing to do here, but some tests for this change would be great! A table test with some different cases seems appropriate.

One concern inline below.

File: lib/eof.go (hunk 1 above)
Comment (pierreca) [reply lib/eof.go:24]: I agree on principle and long-term that's what i'd want too.

To give some context on the string comparison, the code that motivated it is here: 
https://github.com/hashicorp/net-rpc-msgpackrpc/blob/master/client.go#L33

In this case, the error comes in the Response.Error field as a string, and the string is copied into an error object. In the specific case I am seeing the string is: `rpc error: EOF`. 
This error created by this line of code does not unwrap correctly as an EOF and as far as I can tell, cannot be identified as an EOF other than by string comparison. Maybe there's a quick fix I don't know about and if that's the case I'll happily submit another PR for it, but if there isn't i'm afraid the bigger fix involves changing how the RPC client works, and If I understand the situation correctly, this is used in many places in many hashicorp projects. Which makes this change really scary for a first time external contributor. 

Thoughts?

File: lib/eof_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	"fmt"
+	"io"
+	"testing"
+
+	"github.com/hashicorp/yamux"
+	"github.com/stretchr/testify/require"
+)
+
+func TestErrIsEOF(t *testing.T) {
+	t.Parallel()
```
Comment (dnephin) [inline lib/eof_test.go:13]: I know a lot of the tests use `t.Parallel`, but it can cause problems sometimes so I've been trying to remove it where possible. I think in this case, since these tests run so fast, it's actually slower to spin up the goroutine to run the tests in parallel, than it would be to run it without parallel.

```suggestion
```

File: lib/eof_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		err  error
+	}{
+		{name: "EOF", err: io.EOF},
+		{name: "yamuxStreamClosed", err: yamux.ErrStreamClosed},
+		{name: "yamuxSessionShutdown", err: yamux.ErrSessionShutdown},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
```
Comment (dnephin) [inline lib/eof_test.go:26]:
```suggestion
```

File: lib/eof_test.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+			require.True(t, IsErrEOF(tt.err))
+		})
+		t.Run(fmt.Sprintf("Wrapped %s", tt.name), func(t *testing.T) {
+			t.Parallel()
```
Comment (dnephin) [inline lib/eof_test.go:30]:
```suggestion
```

File: lib/eof_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+			require.True(t, IsErrEOF(fmt.Errorf("test: %w", tt.err)))
+		})
+		t.Run(fmt.Sprintf("String suffix is %s", tt.name), func(t *testing.T) {
+			t.Parallel()
```
Comment (dnephin) [inline lib/eof_test.go:34]:
```suggestion
```

File: lib/eof.go (hunk 1 above)
Comment (pierreca) [reply lib/eof.go:24]: I really like this idea, I'll send an update.

The scenario i'm testing is on long-poll requests where a user of the API is connected to a nomad client, the client is forwarding the RPC to a server, and while the request is blocked, the leader is killed. In that specific case, sometimes, I see this wrapped error as string. I think it may have to do with how the requests are forwarded between the client, its server, and potentially the cluster leader. I'm still learning how that codepath works, I've only been staring at it for two days :)

What can I do that would help narrow this down? I can repro on a fairly regular basis, I just need help orienting myself in the code and adding logs where we need them to confirm the problematic code path(s).

Comment (pierreca) [reply lib/eof.go]: I have added the ServerError type assertion: I hope it is restrictive enough and solves at least one of the failure mode I'm trying to get past (explained below, step 3). I'm not out of the woods yet but if this gets checked in, I hope it'll help improve robustness. 

**Long explanation**
I am starting to understand that the problem I'm trying to solve goes deeper and lies in the forwarding chain of the RPC requests in a nomad cluster. Here's the scenario:

1. api user makes a blocking request (wait on an index) -> nomad client -> nomad server (leader)
2. leader dies/is killed while the request is waiting. Returns an EOF error
3. *with this fix*: client retries and changes server: client -> nomad server(follower) (without the fix, the client would not understand it received an EOF error)
4. nomad server (follower) -> nomad server(leader) : OOPS the leader is down: "connection refused" error (not EOF and the leader election hasn't happened yet)
5. follower does not retry and passes the error (connection refused) down to the client in the `response.Error` field
6. client: not an EOF, returns a wrapped version of the connection refused to the caller <- that's my new problem :)

From the caller perspective this is difficult, if not impossible, to handle correctly: the caller shouldn't be expected to understand (and recover from) RPC forwarding chain errors. There are a finite number of scenarios like this (depending whether the client is connected to a follower or a leader, how fast leader election goes, and which server goes down, the one doing the forwarding or the leader). I think this should be entirely handled within the RPC chain between the nomad client and servers. retries and HA are hard problems though :) 

This being said, I don't think it's realistic to expect that all cases can be covered with `IsErrEOF()`, which is already a little bit extended past EOF (the yamuxError, and now some ServerErrors provided they end with `: EOF`). The follow up investigation is probably directly in the nomad RPC code.

In conclusion, I still think this is a net improvement and we've pushed this PR as far as it should go, happy to make more changes as required to get it merged, this is a move in the right direction :)

**Comment by pierreca** [comment]
@dnephin would you mind giving this another look and let me know if these last commits address everything you wanted?

**Review by dnephin — COMMENTED** [review]
I think this implementation is looking good, but I realized I still may not fully understand the specific problem this is fixing. I need to read over your detailed comment above again more carefully to make sure I understand. My main concern is that this `IsErrEOF` function is called from a few places, and I'm not sure if this change will be correct for all of the places.

File: lib/eof.go [hunk 6]
```diff
@@ -23,5 +26,13 @@ func IsErrEOF(err error) bool {
 		return true
 	}
 
+	if srvErr, ok := err.(rpc.ServerError); ok {
+		return strings.HasSuffix(srvErr.Error(), fmt.Sprintf(": %s", io.EOF.Error()))
+	}
+
+	if srvErr, ok := errors.Unwrap(err).(rpc.ServerError); ok {
```
[the remark this answers is not in the document]
Comment (pierreca) [reply lib/eof.go]: Good point. Done.

**Comment by pierreca** [comment]
@dnephin no worries, I understand, i've been on the maintainer side for years in the past. it's hard. happy to jump in a slack conversation or something like that if it makes it easier.

The bottom line is that I have found that the nomad retry logic is broken in rpc forwarding scenarios and that part of the problem is not all EOF errors aren't properly identified as such. What this change does is widen the scope of what gets recognized as `EOF` to include wrapped `io.EOF` errors and EOF errors wrapped in `ServerError` that are sent in the RPC response body. 

What are the risks?
- interpreting an error that isn't an EOF as an EOF: this was your original concern, I believe it's addressed by narrowing it down to precise type checking
- kicking in retries where the code would not retry before, which has 2 possibilities: 
    - it should've retried (that's a bugfix then, and that's what I'm going for and seeing in my tests)
    - it shouldn't have retried: that'd be either a bad interpretation of the error or a really weird case (ok to retry on EOF but not wrapped EOF is a weird scenario IMHO). It should be fairly easy to review those few cases just by looking at where `IsErrEOF` is used:
      - https://github.com/hashicorp/consul/blob/master/agent/pool/pool.go#L609
      - https://github.com/hashicorp/consul/blob/master/agent/consul/rpc.go#L504
      - https://github.com/hashicorp/nomad/blob/master/helper/pool/pool.go#L460
      - https://github.com/hashicorp/nomad/blob/master/client/rpc.go#L111

IDK what your test coverage looks like but if there are tests I can write to help you make sure this doesn't break, just point me where. I think there is one or two more changes needed in nomad after that to fix retries in RPC forwarding chains and that'd be a win for a lot of users :)

**Comment by pierreca** [comment]
@dnephin ping? anything else I can do to help get this checked in? we hit errors related to nomad not retrying when it should very consistently.

**Comment by pierreca** [comment]
@dnephin still hoping to upstream this, lmk if I can help address any doubt that this is a good fix.

File: lib/eof_test.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	var tests = []struct {
+		name string
+		err  error
+	}{
+		{name: "EOF", err: io.EOF},
+		{name: "Wrapped EOF", err: fmt.Errorf("test: %w", io.EOF)},
+		{name: "yamuxStreamClosed", err: yamux.ErrStreamClosed},
+		{name: "yamuxSessionShutdown", err: yamux.ErrSessionShutdown},
+		{name: "ServerError(___: EOF)", err: rpc.ServerError(fmt.Sprintf("rpc error: %s", io.EOF.Error()))},
+		{name: "Wrapped ServerError(___: EOF)", err: fmt.Errorf("rpc error: %w", rpc.ServerError(fmt.Sprintf("rpc error: %s", io.EOF.Error())))},
```
Comment (dnephin) [inline lib/eof_test.go:23]: A couple tests that are expected to return false might be good as well, to check that errors which are not EOF are not matched.

**Review by dnephin — APPROVED** [review]
Thank you for the fix!

One suggestion for some additional test cases, but not blocking.
